base64ct/src/lib.rs

```diff
@@ -417,3 +392,36 @@ fn unpadded_len_ct(input: &[u8]) -> usize {
         _ => input.len(),
     }
 }
+
+/// Validate padding is well-formed.
+///
+/// Returns length-related errors eagerly as a [`Result`]], and data-dependent
+/// errors (i.e. malformed padding bytes) as `i16` to be combined with other
+/// encoding-related errors prior to branching.
+fn validate_padding(input: &[u8]) -> Result<i16, DecodeError> {
+    if input.len() % 4 != 0 {
+        match input.last().cloned() {
+            // TODO(tarcieri): constant-time whitespace check
+            Some(byte) if char::from(byte).is_whitespace() => {
+                return Err(DecodeError::InvalidEncoding)
+            }
+            _ => return Err(DecodeError::InvalidLength),
```

Comment:
Not necessarily sure this is a good idea. In fact it seems like the sort of wiggle room that allows things like padding oracle attacks (not saying padding oracle attacks are applicable here)

Perhaps we should just have `EncodeError` and `DecodeError`, or a unit `Error` type?

Reply:
I think it should be enough to simply return `InvalidEncoding`, since incorrect string length is an encoding problem, not buffer length one.

Reply:
That's a potential option, although note that in addition to `InvalidLength` being a potentially confusing error to get for trailing whitespace, it causes a discrepancy between the error returned for trailing whitespace between the padded and unpadded decoding functions, at least given the current implementation.

Reply:
(ignore my previous reply, I was confused)

Done. This seems ok for now.